File: Day1/main.py
import re

SAMPLE_DATA: bool = False
if SAMPLE_DATA:
    filename = "sample_data.txt"
else:
    filename = "data.txt"

patt = re.compile(r'((?:\d+\n?)+)\n?', re.MULTILINE)

with open(filename, 'r') as f:
    calories = sorted([sum([int(s) for s in m.strip().split('\n')]) for m in patt.findall(f.read())], reverse=True)[:3]
print(f'{calories[0]} {sum(calories)}')

File: Day1/main.c
#include <stdio.h>

#if 0
#define FILENAME "sample_data.txt"
#else
#define FILENAME "data.txt"
#endif


int main() {
    FILE *fp = fopen(FILENAME, "r");
    if (fp == NULL) {
        printf("Error opening file " FILENAME "\n");
        return 1;
    }
    int top_three[3] = {-1, -1, -1};

    char line[256];

    int value = 0;

    while (fgets(line, sizeof(line), fp)) {
        if (line[0] == '\n'){
            value = 0;
            continue;
        }
        value += atoi(line);
        for (int i = 0; i < 3; i++) {
            if (value > top_three[i]) {
                for (int j = 2; j >= i; j--) {
                    top_three[j] = top_three[j - 1];
                }
                top_three[i] = value;
                break;
            }
        }
    }
    int sum = 0;
    for (int i = 0; i < 3; i++) {
        sum += top_three[i];
    }

    printf("Sum of top three: %d\n", sum);
    printf("Max: %d\n", top_three[0]);

    return 0;
}